runtime/protocol/http-client/common/src/aws/smithy/kotlin/runtime/http/interceptors/FlexibleChecksumsRequestInterceptor.kt

```diff
@@ -7,102 +7,135 @@ package aws.smithy.kotlin.runtime.http.interceptors
+ * If a user supplies a checksum via an HTTP header no calculation will be done. The exception is MD5, if a user
+ * supplies an MD5 checksum header it will be ignored.
  *
- * Otherwise, if it will be sent as a trailing header, calculate the checksum as asynchronously as the body is streamed.
- * In this case, a [LazyAsyncValue] will be added to the execution context which allows the trailing checksum to be sent
- * after the entire body has been streamed.
+ * If the request configuration and model requires checksum calculation:
+ * - Check if the user configured a checksum algorithm for the request and attempt to use that.
+ * - If no checksum is configured for the request then use the default checksum algorithm to calculate a checksum.
  *
- * @param checksumAlgorithmNameInitializer an optional function which parses the input [I] to return the checksum algorithm name.
- * if not set, then the [HttpOperationContext.ChecksumAlgorithm] execution context attribute will be used.
+ * If the request will be streamed:
+ * - The checksum calculation is done during transmission using a hashing & completing body.
+ * - The checksum will be sent in a trailing header, once the request is consumed.
+ *
+ * If the request will not be streamed:
+ * - The checksum calculation is done before transmission
+ * - The checksum will be sent in a header
+ *
+ * Business metrics MUST be emitted for the checksum algorithm used.
+ *
+ * @param requestChecksumRequired Model sourced flag indicating if checksum calculation is mandatory.
+ * @param requestChecksumCalculation Configuration option that determines when checksum calculation should be done.
+ * @param requestChecksumAlgorithm The checksum algorithm that the user selected for the request, may be null.
  */
 @InternalApi
 public class FlexibleChecksumsRequestInterceptor<I>(
-    private val checksumAlgorithmNameInitializer: ((I) -> String?)? = null,
+    requestChecksumRequired: Boolean,
+    requestChecksumCalculation: HttpChecksumConfigOption?,
+    requestChecksumAlgorithm: String?,
 ) : AbstractChecksumInterceptor() {
-    private var checksumAlgorithmName: String? = null
 
-    @Deprecated("readAfterSerialization is no longer used")
+    // FIXME: Remove in next minor version bump
+    @Deprecated("Old constructor is no longer used but it's kept for backwards compatibility")
+    public constructor() : this(
+        false,
+        HttpChecksumConfigOption.WHEN_REQUIRED,
+        null,
+    )
+
+    // FIXME: Remove in next minor version bump
+    @Deprecated("Old constructor is no longer used but it's kept for backwards compatibility")
+    public constructor(
+        checksumAlgorithmNameInitializer: ((I) -> String?)? = null,
+    ) : this(
+        false,
+        HttpChecksumConfigOption.WHEN_REQUIRED,
+        null,
+    )
+
+    private val checksumHeader = buildString {
+        append("x-amz-checksum-")
+        append(requestChecksumAlgorithm?.lowercase() ?: "crc32")
```

Comment:
style: "crc32" could be lifted to a `const val SDK_DEFAULT_CHECKSUM_ALGORITHM` for easier understanding